#Tugas Lab 2
Nama: Muhammad Nur Aziz

Kelas: 3F

NIM: 2341720237


Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur CountVectorizer dengan mengaktifkan stop_words

3. Evaluasi hasilnya

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur TF-IDF dengan mengaktifkan stop_words

3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [35]:
from IPython.display import clear_output
!pip install gdown
clear_output()
!gdown --id "13NKYtYtLpL6FL9HgEEIzQntcC03wt2p7"

/usr/local/lib/python3.12/dist-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=13NKYtYtLpL6FL9HgEEIzQntcC03wt2p7
To: /content/spam.csv
100% 504k/504k [00:00<00:00, 8.73MB/s]


In [36]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.naive_bayes import MultinomialNB

In [46]:
df = pd.read_csv('spam.csv', encoding='latin-1')
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [47]:
df = df[['v1', 'v2']]
df.columns = ['label', 'message']

In [48]:
#slit
df['label'] = df['label'].map({'ham': 0, 'spam': 1})

X = df['message']
y = df['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#1. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur CountVectorizer dengan mengaktifkan stop_words

3. Evaluasi hasilnya

In [49]:
# CountVectorizer stop words
count_vectorizer = CountVectorizer(stop_words='english')
X_train_count = count_vectorizer.fit_transform(X_train)
X_test_count = count_vectorizer.transform(X_test)

In [50]:
# Model Naive Bayes dengan CountVectorizer
nb_count = MultinomialNB()
nb_count.fit(X_train_count, y_train)
y_pred_count = nb_count.predict(X_test_count)

In [51]:
accuracy_count = accuracy_score(y_test, y_pred_count)
print(f"Akurasi CountVectorizer: {accuracy_count:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_count))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_count))

Akurasi CountVectorizer: 0.9839

Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Confusion Matrix:
[[959   6]
 [ 12 138]]


#2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur TF-IDF dengan mengaktifkan stop_words

3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [54]:
# TF-IDF stop words
tfidf_vectorizer = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

In [55]:
# Model Naive Bayes dengan TF-IDF
nb_tfidf = MultinomialNB()
nb_tfidf.fit(X_train_tfidf, y_train)
y_pred_tfidf = nb_tfidf.predict(X_test_tfidf)

In [56]:
accuracy_tfidf = accuracy_score(y_test, y_pred_tfidf)
print(f"Akurasi TF-IDF: {accuracy_tfidf:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_tfidf))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_tfidf))

Akurasi TF-IDF: 0.9668

Classification Report:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115


Confusion Matrix:
[[965   0]
 [ 37 113]]


In [57]:
print(f"Akurasi CountVectorizer: {accuracy_count:.4f}")
print(f"Akurasi TF-IDF: {accuracy_tfidf:.4f}")
print(f"Selisih Akurasi: {abs(accuracy_count - accuracy_tfidf):.4f}")

Akurasi CountVectorizer: 0.9839
Akurasi TF-IDF: 0.9668
Selisih Akurasi: 0.0170


#Kesimuplan
- CountVectorizer menunjukkan performa yang lebih baik dari pada TF-IDF

- CountVectorizer mampu mendeteksi  lebih banyak recall pesan spam yang sebesar 92% dibanding TF-IDF yang sebesear 75%
